<a href="https://colab.research.google.com/github/map35/Coursera_Capstone/blob/main/Toronto_Neighborhoods.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Library**

In [1]:
pip install geocoder

     |████████████████████████████████| 98 kB 3.8 MB/s 


In [2]:
import numpy as np
import pandas as pd
import matplotlib.cm as cm
import matplotlib.colors as colors
import requests
import json
import folium
import geocoder
from sklearn.cluster import KMeans
from pandas.io.json import json_normalize
from geopy.geocoders import Nominatim
from bs4 import BeautifulSoup

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# **Scrapping Toronto Neighborhood Data From Wikipedia**

In [3]:
# Making soup object with the URL
response = requests.get(url="https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M",)
soup = BeautifulSoup(response.content, 'html.parser')

# Making List of table content
table_contents = []
table = soup.find('table')
for row in table.findAll('td'):
    cell = {}
    # Exclude borough with value 'Not assigned'
    if row.span.text == 'Not assigned':
        pass
    else:
        cell['Postal Code'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# Transform the List into DataFrame
df = pd.DataFrame(table_contents)
df['Borough'] = df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
# Checking our DataFrame
df.head()

Postal Code           Borough                      Neighborhood
0         M3A        North York                         Parkwoods
1         M4A        North York                  Victoria Village
2         M5A  Downtown Toronto         Regent Park, Harbourfront
3         M6A        North York  Lawrence Manor, Lawrence Heights
4         M7A      Queen's Park     Ontario Provincial Government

In [4]:
# Checking the shape of our DataFrame
df.shape

(103, 3)

There are 103 rows in the Toronto Neighborhood DataFrame

# **Explore Neighborhood in Toronto**

Upload Geospatial_Coordinates.csv

In [5]:
coor = pd.read_csv("https://raw.githubusercontent.com/map35/Coursera_Capstone/main/Geospatial_Coordinates.csv")
coor.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Merged the neighborhood dataframe with the coordinates 

In [6]:
toronto_df = df
toronto_df = toronto_df.join(coor.set_index('Postal Code'), on='Postal Code')

toronto_df.head()

Postal Code           Borough                      Neighborhood   Latitude  \
0         M3A        North York                         Parkwoods  43.753259   
1         M4A        North York                  Victoria Village  43.725882   
2         M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3         M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4         M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

Make Toronto map

In [7]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [8]:
toronto_map = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Borough'], toronto_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(toronto_map)  
    
toronto_map

Making function to get venue in Toronto

In [9]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            'X1SDU4HLNBGV2IMYFSXPJIWBDLQUO2M50Z2ZNCY4Q0NVF3IY', 
            'VOYRHOJHPKHC5UXSXTVFIYO5BZQC4HTUO0UICSPBYYV3CQFE', 
            20180605, 
            lat, 
            lng, 
            500, 
            100)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [10]:
toronto_venues = getNearbyVenues(names=toronto_df['Neighborhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [11]:
print(toronto_venues.shape)
toronto_venues.head()

(2136, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1                     KFC       43.754387       -79.333021   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4             Tim Hortons       43.725517       -79.313103   

         Venue Category  
0                  Park  
1  Fast Food Restaurant  
2     Food & Drink Shop  
3          Hockey Arena  
4           Coffee Shop

In [12]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Alderwood, Long Branch                                                  7   
Bathurst Manor, Wilson Heights, Downsview North                        23   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      24   
Berczy Park                                                            58   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           23   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Caledonia-Fairbanks                                                     4   
Cedarbrae                                                               9   
Central Bay Street                                                     67   
Christie                                                               15   
Church and Wellesley                                                   78   
Clairville, Humberwood, Woodbine Downs, West Hu...                      4   
Clarks Corners, Tam O'Shanter, Sullivan                                14   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        8   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        5   
Don Mills North                                                         6   
Don Mills South                                                        19   
Dorset Park, Wexford Heights, Scarborough Town ...                      6   
Downsview Central                                                       3   
Downsview East                                                          2   
Downsview Northwest                                                     4   
Downsview West                                                          5   
Dufferin, Dovercourt Village                                           17   
Enclave of L4W                                                         12   
Enclave of M4L                                                         16   
Enclave of M5E                                                         97   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                     10   
Fairview, Henry Farm, Oriole                                           64   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Glencairn                                                               4   
Golden Mile, Clairlea, Oakridge                                         8   
Guildwood, Morningside, West Hill                                       8   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
Hillcrest Village                                                       5   
Humber Summit                                                           1   
Humberlea, Emery                                                        1   
Humewood-Cedarvale                                                      4   
India Bazaar, The Beaches West                                         19   
Kennedy Park, Ionview, East Birchmount Park                             5   
Kensington Market, Chinatown, Grange Park                              71   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                

In [13]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 276 uniques categories.


# **Analyze Neighborhood in Toronto**



In [14]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Arepa Restaurant  Art Gallery  Art Museum  \
0             0         0                 0            0           0   
1             0         0                 0            0           0   
2             0         0                 0            0           0   
3             0         0                 0            0           0   
4             0         0                 0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Bowling Alley  \
0       0              0          0         0              0   
1       0              0          0         0              0   
2       0              0          0         0              0   
3       0              0          0         0              0   
4       0              0          0         0              0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0    

In [15]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                           Agincourt     0.000000   
1                              Alderwood, Long Branch     0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North     0.000000   
3                                     Bayview Village     0.000000   
4                   Bedford Park, Lawrence Manor East     0.000000   
5                                         Berczy Park     0.000000   
6                         Birch Cliff, Cliffside West     0.000000   
7        Brockton, Parkdale Village, Exhibition Place     0.000000   
8   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
9                                 Caledonia-Fairbanks     0.000000   
10                                          Cedarbrae     0.000000   
11                                 Central Bay Street     0.014925   
12                                           Christie     0.000000   
13                               Church and Wellesley     0.025641   
14  Clairville, Humberwood, Woodbine Downs, West H...     0.000000   
15            Clarks Corners, Tam O'Shanter, Sullivan     0.000000   
16    Cliffside, Cliffcrest, Scarborough Village West     0.000000   
17                     Commerce Court, Victoria Hotel     0.010000   
18                                         Davisville     0.000000   
19                                   Davisville North     0.000000   
20   Del Ray, Mount Dennis, Keelsdale and Silverthorn     0.000000   
21                                    Don Mills North     0.000000   
22                                    Don Mills South     0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...     0.000000   
24                                  Downsview Central     0.000000   
25                                     Downsview East     0.000000   
26                                Downsview Northwest     0.000000   
27                                     Downsview West     0.000000   
28                       Dufferin, Dovercourt Village     0.000000   
29                                     Enclave of L4W     0.000000   
30                                     Enclave of M4L     0.000000   
31                                     Enclave of M5E     0.010309   
32  Eringate, Bloordale Gardens, Old Burnhamthorpe...     0.000000   
33                       Fairview, Henry Farm, Oriole     0.000000   
34             First Canadian Place, Underground city     0.000000   
35                           Forest Hill North & West     0.000000   
36                           Garden District, Ryerson     0.000000   
37                                          Glencairn     0.000000   
38                    Golden Mile, Clairlea, Oakridge     0.000000   
39                  Guildwood, Morningside, West Hill     0.000000   
40  Harbourfront East, Union Station, Toronto Islands     0.000000   
41                      High Park, The Junction South     0.000000   
42                                  Hillcrest Village     0.000000   
43                                      Humber Summit     0.000000   
44                                   Humberlea, Emery     0.000000   
45                                 Humewood-Cedarvale     0.000000   
46                     India Bazaar, The Beaches West     0.000000   
47        Kennedy Park, Ionview, East Birchmount Park     0.000000   
48          Kensington Market, Chinatown, Grange Park     0.000000   
49  Kingsview Village, St. Phillips, Martin Grove ...     0.000000   
50                   Lawrence Manor, Lawrence Heights     0.000000   
51                                      Lawrence Park     0.000000   
52                                            Leaside     0.000000   
53                           Little Portugal, Trinity     0.022727   
54                                     Malvern, Rouge     0.000000   
55  Milliken, Agincourt North, Steeles East, L'Amo...     0.000000   
56  Mimico NW, The Queensway West, South of Bloor,.

In [17]:
toronto_grouped.shape

(99, 276)

In [20]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [21]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village    Chinese Restaurant   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

       2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant               Skating Rink        Breakfast Spot   
1                        Gym             Sandwich Place          Dance Studio   
2                       Bank  Middle Eastern Restaurant     Convenience Store   
3                       Café                       Bank   Japanese Restaurant   
4                Coffee Shop             Sandwich Place                   Pub   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0      Doner Restaurant                 Diner        Discount Store   
1                   Pub           Coffee Shop        Airport Lounge   
2           Bridal Shop        Sandwich Place                 Diner   
3         Women's Store                 Diner        Discount Store   
4   Japanese Restaurant          Liquor Store       Thai Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Distribution Center              Dive Bar                Dog Run  
1      Department Store           Escape Room      Electronics Store  
2            Restaurant          Intersection          Deli / Bodega  
3   Distribution Center              Dive Bar                Dog Run  
4  Fast Food Restaurant     Indian Restaurant            Pizza Place

In [22]:
neighborhoods_venues_sorted.shape

(99, 11)